pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java

```diff
@@ -145,9 +145,10 @@ public int filterEntriesForConsumer(Optional<EntryWrapper[]> entryWrapper, int e
             MessageMetadata msgMetadata = entryWrapper.isPresent() && entryWrapper.get()[entryWrapperIndex] != null
                     ? entryWrapper.get()[entryWrapperIndex].getMetadata()
                     : null;
-            msgMetadata = msgMetadata == null
-                    ? Commands.peekMessageMetadata(metadataAndPayload, subscription.toString(), -1)
-                    : msgMetadata;
+            if (msgMetadata == null) {
```

Comment:
Looks like we will not share `msgMetadata` to other threads? I have checked where used the `msgMetadata`, they should run in the same thread. Maybe I missed something.

It's a good motivation, is it better to avoid passing the `msgMetadata`? Instead, we can create local variable, looks like

```
boolean hasTxnidMostBits = msgMetadata.hasTxnidMostBits();
int batchSize = msgMetadata.getNumMessagesInBatch();
msgMetadata.clear();

// The followings should use the `hasTxnidMostBits` and `batchSize` directly.
```

Reply:
> It's a good motivation, is it better to avoid passing the msgMetadata? Instead, we can create local variable, looks like

yes, avoiding the instance is a better choice. I'll take a look in the refactoring.

Reply:
I took a look in the possible refactoring. there's quite a few fields that are used from MessageMetadata, so it would turn out to be an ugly solution. 
It's possible that a copy isn't necessary in this case, but it's hard to tell without checking all accesses.

Reply:
> If we have to pass it to another method, we should make a copy.

I guess that would be one way to define a rule when to make a copy. (btw. I wonder if it's intentional that the generated copyFrom code doesn't call "clear" as the first step. )